Replace sass with Postcss

occupant opened the pull request:
Not sure if this is something we need to do necessarily, but this is an example of what it would look like to replicate the current sass functionality with postcss. The sass files themselves remain essentially unchanged (other than a few comment format changes and moving imports to the top of the file).

I've also altered the ckeditor5.css file, removing the prefix class as that can now be done via the postcss.config.js. 

I think in general the sass variables would be better as css variables, the nesting isn't necessarily needed (native nesting is coming although I don't use it so am not sure where support is atm) and the minification is probably unnecessary, but that's all up for discussion.

The one thing that we do lose is the mixins. There may be a way to support those as well, but I'm not convinced we actually need those.

For now, I've set the css to output into the dist/css/ directory so the output can be reviewed.

joelpittet commented:
I 100% agree we don't need nesting, it causes more problems than it solves, when trying to find selectors in your sass from a CSS on the page it can be painful (without maps)

joelpittet commented:
Also minification is terrible because it makes it hard to see diffs, let Drupal aggregation deal with that, so 100% agree there too

joelpittet commented:
CSS variables is a toss up... but I can get onboard with that.

joelpittet commented on package.json:
```diff
@@ -1,10 +1,18 @@
   "scripts": {
-    "build:css": "sass sass:css",
-    "watch:css": "sass -w sass:css"
+    "watch": "NODE_ENV=production postcss ./sass/*.scss -d ./css --ext css -w",
+    "css": "NODE_ENV=production postcss ./sass/*.scss -d ./css --ext css"
```
Can we go with something like this without the the `NODE_ENV` and move `sass` to `src/css`?
```
    "build:css": "postcss src/css/*.css -o css/main.css",
    "watch:css": "postcss src/css/*.css -o css/main.css --watch"
```

joelpittet commented on package.json:
```diff
@@ -1,10 +1,18 @@
   },
   "devDependencies": {
-    "sass": "^1.0.0"
+    "autoprefixer": "^10.4.13",
+    "cssnano": "^5.1.15",
+    "postcss": "^8.4.21",
+    "postcss-cli": "^9.1.0",
+    "postcss-import": "^14.1.0",
+    "postcss-loader": "^6.2.1",
+    "postcss-nesting": "^13.0.0",
+    "postcss-prefix-selector": "^1.16.1",
+    "postcss-simple-vars": "^6.0.3"
```
This seems like a lot, do we need all of those?
How about just these to keep the dependencies light?
```
    "postcss": "^8.0.0",
    "postcss-cli": "^9.0.0",
    "autoprefixer": "^10.0.0",
    "postcss-prefix-selector": "^1.0.0",
    "postcss-preset-env": "^7.0.0"
```

occupant replied:
Yes, agreed. `cssnano` and `postcss-loader` can definitely go (didn't mean to include the loader). 

Not a fan of nesting whether natively supported or not, so `postcss-nesting` can for sure go.

`postcss-simple-vars` can easily be replaced by css native variables, so no need for that.

`postcss-import` is really only useful if you break out your css files into components or other subsets, which I don't think is needed.

So I'm good with all of those suggestions.

joelpittet commented on postcss.config.js:
```diff
@@ -0,0 +1,14 @@
+module.exports = {
+  plugins: [
+    require("postcss-import"),
+    require("postcss-simple-vars"),
+    require("postcss-nesting"),
+    require("postcss-prefix-selector")({
+      prefix: ".ck-content",
+      exclude: ["body", "html", ":root"],
+      includeFiles: "sass/ckeditor5.scss",
+    }),
+    require("autoprefixer"),
+    //...(process.env.NODE_ENV === "production" ? [require("cssnano")] : []),
+  ],
+};
```
If we have less dependencies this might look like this.
```
module.exports = {
  plugins: [
    require('postcss-preset-env')({
      stage: 1, // Enables future CSS features at stage 1
      features: {
        'nesting-rules': true // Enable CSS nesting (similar to what SASS offers)
      }
    }),
    require("postcss-prefix-selector")({
      prefix: ".ck-content",
      exclude: ["body", "html", ":root"],
      includeFiles: "sass/ckeditor5.scss",
    }),
    require('autoprefixer') // Automatically adds vendor prefixes based on browser support
  ]
};
```

joelpittet reviewed (changes requested):
I think this is safe enough to try, I just want to try to keep the dependencies as light as possible, so I even question the need for the prefixing one... what's you're take on that?

occupant commented:
> I think this is safe enough to try, I just want to try to keep the dependencies as light as possible, so I even question the need for the prefixing one... what's you're take on that?

In terms of prefixing:

I find this really useful, although I'm using it slightly differently, so may not be entirely relevant here:
I use it to prefix my main stylesheet and then load that in the editor so that front-end styles are reflected without worrying about style bleed into the admin theme.

If that's not something we need / want here, we don't need a node package to write a few prefixes in the ckeditor-specific stylesheet.

occupant commented:
With the latest commits, I've 

- updated the node packages
- moved / renamed the sass files to src/css
- removed all sass variables (didn't replace with css variables yet as most won't be needed and should have naming convention / utility if created)
- removed all nesting (even though enabled, not sure it's something we need)
- recompiled

I guess the question now is: what are we getting from node? Autoprefixer, nesting, the watch task, and, maybe, the automated prefix prepending. Is that enough to even require a build step? I'm not sure it is. The more node tasks we remove, the less convinced I am that what remains is adding value. Anything else we'd be losing out on if we just moved to pure css?

joelpittet commented:
@occupant we might lose our seat at the cool kids table ;)

I am game to ditch node in this theme if you are!

occupant commented:
@joelpittet Ok, let's do it! I'll meet you at the wobbly table in the corner. 

I'll close the PR and we can make another without node at all.
